imodeljs/imodeljs
# Deliver ES Modules

**tl;dr**: Deliver ESModules where possible

Here's hopefully the last major breaking change for 3.0. This is mainly being done so we can take advantage of dynamic imports to improve our bundle sizes and a pre-req for the new Extension API. All frontend and common pkgs will deliver both cjs and esm, while backend pkgs will continue to just deliver cjs as we don't expect to see better esm support in node for a while.

- Updated @bentley/build-tools/tsconfig-base.json
	- Bumped [target](https://www.typescriptlang.org/tsconfig#target) from [es2017](https://flaviocopes.com/es2017/) -> [es2019](https://flaviocopes.com/es2019/)
		- take a look at [compatibility](https://kangax.github.io/compat-table/es2016plus/)
		- initially hoped to target es2020, but webpack had trouble parsing some newer operators like optional chaining and nullish coalescing, so we'll target es2019 and still be able to use those features, but we'll have ts transpile it to something webpack is able to parse

- All packages that we currently ship deliver their cjs build to `lib` but will now be delivered to `lib/cjs`
	- This really shouldn't affect users unless they were importing directly from lib, in which case they'll need to update the import path from `@bentley/foo/lib/bar` to `@bentley/foo/lib/cjs/bar`
		- The only place we, core, import from lib is typically when a package delivers both a frontend + backend module in the same package, ie ElectronHost from the electron-manager package; or in tests which I'll discuss later 	
- All frontend packages will now be delivering an esm build to `lib/esm`

- Presentation:
	- The presentation pkgs have a lot of internal testing helpers that are used amongst each other. I added a barrel file for tests so we can clean up all the imports. I originally re-exported everything from the test barrel in the pkg's main barrel in hopes of completely removing imports from lib. Although the cjs and esm types are exactly the same, typescript was throwing an error that they didn't match when I set the package.json typings field to point to types generated by the esm build, there was also an issue with some of the statics so i reverted the change of exporting test-helpers in the main barrel, but kept the test barrel
	- presentation-common was re-exporting the entire typemoq pkg and this was being used in tests in ui-framework and was just wrong imo, so that's been removed

- UI Pkgs:
	- ui-abstract: rename files with `*.tsx` file extension to `*.ts` as this pkg should be ui-framework agnostic
	- most of the changes in the other ui-pkgs are to react to where we deliver scss
		- for example in the pkgs that use scss, we typically import scss from lib directly like so, 
			    ```
			    @import "~@bentley/ui-core/lib/ui-core/icons/variables";
			    ```
			  so had to change it to:
			    ```
			    @import "~@bentley/ui-core/lib/cjs/ui-core/icons/variables";
			    ```
		- in my opinion we should split delivering scss out from these pkgs and deliver a scss only pkg like iTwinUi does

- Test Apps:
	- Verified Display and Presentation Test App work. 
	- Ui-Test-App is currently broken, as is on master due to the recent auth changes.
 
- General comments re testing: 
	- Many of our pkgs don't export internals or testing utilities in the main barrel file of that pkg so we have to import from lib directly. We actually never deliver these files, as they're ignored in the `.npmignore` so there is no way anyone outside the monorepo from being able to write their own tests using those test utilities, and I guess they really shouldn't but who are we to stop them. I'd recommend pulling those out into individual testing pkgs and consume those accordingly. 

- For consumers:
	- This PR should hopefully drastically decrease web app bundles as any unnecessary code should now be tree-shaken away. (Once we disable the current IModeljsExportPlugin in our CRA config used for the 2.x extension api)
	- Webpack will default to prioritizing `esm` over `cjs`, but this can be overriden with the [resolve.mainFields](https://webpack.js.org/configuration/resolve/#resolvemainfields) 


Some further reading for those interested:
- https://web.dev/commonjs-larger-bundles/

To do in a following PR: 
- ~~merge master (with the massive rename pr) in 😭~~
- update NextVersion.md (@aruniverse)
- try to update scss and public asset delivery (@aruniverse)
- update to use built in v8 coverage to replace disable istanbul-instrumented loader (@wgoehrig)
- Revert change made to function coverage in presentation-components (@saskliutas)
